keps/sig-storage/3762-persistent-volume-last-phase-transition-time/README.md

```diff
@@ -0,0 +1,907 @@
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+We need to update API server to support the newly proposed field and update PV controller to set a value of the new
+timestamp field when a volume transitions to a different phase. Also, if the feature gate is disabled the value must be
+re-set to `nil` when updating or creating a volume.
+
+The value of the field is not intended for use by any other Kubernetes components at this point and should be used only
+as a convenience feature for cluster admins. Cluster admins should be able to list and sort PersistentVolumes based on
+a timestamp which indicates when the volume transitioned to a different state. 
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+As a cluster admin I want to use `Retain` policy for released volumes, which is safer than `Delete`, and implement a
+reliable policy to delete volumes that are `Released` for more than X days.
+
+#### Story 2
+
+As a cluster admin I want to be able to reason about volume deletion, or produce alerts, based on a volume being in
+`Pending` phase for more than X hours.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+The new field is purely informative and should not introduce any risk.
+
+## Design Details
+
+Changes required for this KEP:
+
+* kube-apiserver
+  * extend [PersistentVolumeStatus](https://github.com/kubernetes/api/blob/a26a16a095cab454e928a95c533b8cf1b80aa2ec/core/v1/types.go#L402) type with `LastPhaseTransitionTime` field:
+  ```
+    type PersistentVolumeStatus struct {
+    ...
+    // lastPhaseTransitionTime represents a point in time as a timestamp of when a volume last transitioned its phase.
+    // +optional
+    LastPhaseTransitionTime string `json:"lastPhaseTransitionTime,omitempty" protobuf:"bytes,4,opt,name=lastPhaseTransitionTime"`
```

Comment:
I contemplated a bit on how useful it would be to have a history of all the phase transitions, ie available -> bound -> released.

But I think it would end up trying to fit retrofit the legacy phases into conditions, which may be a bit overkill for this use case.

Reply:
That certainly could provide more insight into volume lifecycle, although it's probably not needed too often. And while not as elegant as having conditions the history of a volume and it's phases should be traceable from pv controller [logs](https://github.com/kubernetes/kubernetes/blob/7a55b76f28eddbbb7abf69038d4bd5abab833b4f/pkg/controller/volume/persistentvolume/pv_controller.go#L851).